File: sdk/node/src/verify.ts
// src/verify.ts
// Webhook signature verification for Node.js.
//
// Usage (Express):
//   import { verifyWebhook } from 'agentguard'
//
//   app.post('/webhooks/agentguard', express.raw({type: '*/*'}),
//     (req, res) => {
//       const valid = verifyWebhook(
//         req.body,
//         req.headers['x-agentguard-signature'] as string,
//         process.env.AGENTGUARD_WEBHOOK_SECRET!
//       )
//       if (!valid) return res.status(403).send('Forbidden')
//       const payload = JSON.parse(req.body)
//       // handle...
//     }
//   )

import { createHmac, timingSafeEqual } from 'crypto'

/**
 * Verify an AgentGuard webhook signature.
 *
 * @param body      - Raw request body (Buffer or string)
 * @param signature - Value of X-AgentGuard-Signature header
 * @param secret    - Your AGENTGUARD_WEBHOOK_SECRET
 * @returns         - true if signature is valid
 */
export function verifyWebhook(
  body:      Buffer | string,
  signature: string,
  secret:    string,
): boolean {
  if (!signature || !secret) return false

  // Strip "sha256=" prefix
  const sig = signature.startsWith('sha256=')
    ? signature.slice(7)
    : signature

  const expected = createHmac('sha256', secret)
    .update(body)
    .digest('hex')

  // Timing-safe comparison
  try {
    return timingSafeEqual(
      Buffer.from(expected, 'hex'),
      Buffer.from(sig,      'hex'),
    )
  } catch {
    // Buffer lengths differ = definitely invalid
    return false
  }
}

File: sdk/python/agentguard/verify.py
# agentguard/verify.py
# Webhook signature verification.
# When AgentGuard delivers a webhook to your URL,
# it includes an X-AgentGuard-Signature: sha256=<hex> header.
# Use verify_webhook() to confirm the payload is authentic.
# Usage (FastAPI example):
#   from agentguard import verify_webhook
# @app.post("/webhooks/agentguard")
#   async def handle_escalation(
#       request: Request,
#       x_agentguard_signature: str = Header(None)
#   ):
#       body = await request.body()
#       if not verify_webhook(body, x_agentguard_signature, SECRET):
#           raise HTTPException(403, "Invalid signature")
#       payload = await request.json()
# handle payload...

import hashlib
import hmac


def verify_webhook(
    body:      bytes,
    signature: str,
    secret:    str,
) -> bool:
    """
    Verify an AgentGuard webhook signature.

    Args:
        body:      Raw request body bytes (before JSON parsing)
        signature: Value of X-AgentGuard-Signature header
        secret:    Your AGENTGUARD_WEBHOOK_SECRET

    Returns:
        bool: True if signature is valid, False otherwise.

    Security:
        Uses timing-safe comparison to prevent timing attacks.
    """
    if not signature or not secret:
        return False

    # Strip "sha256=" prefix
    if signature.startswith("sha256="):
        signature = signature[7:]

    expected = hmac.new(
        secret.encode("utf-8"),
        body,
        hashlib.sha256,
    ).hexdigest()

    # Timing-safe comparison
    return hmac.compare_digest(expected, signature)
